tonic/src/codec/compression.rs

```diff
@@ -175,6 +214,16 @@ pub(crate) fn compress(
 
             std::io::copy(&mut gzip_encoder, &mut out_writer)?;
         }
+        #[cfg(feature = "zstd")]
+        CompressionEncoding::Zstd => {
+            let out_writer = bytes::BufMut::writer(out_buf);
+            zstd::stream::copy_encode(
```

Comment:
Wouldn't it be nicer to just mirror the gzip approach and use `zstd::stream::read::{Decoder, Encoder}` instead such that the code can look almost identical:

```
let mut zstd_encoder = Encoder::new(
    &decompressed_buf[0..len],
    // FIXME: support customizing the compression level
    6,
)?;
let mut out_writer = bytes::BufMut::writer(out_buf);
std::io::copy(&mut zstd_encoder, &mut out_writer)?;
```

and decompression:

```
let mut zstd_decoder = Decoder::new(&compressed_buf[0..len])?;
let mut out_writer = bytes::BufMut::writer(out_buf);

std::io::copy(&mut zstd_decoder, &mut out_writer)?;
```

Reply:
@mpetri I was able to add your changes for `decompress` in [384e95e](https://github.com/hyperium/tonic/pull/1371/commits/384e95ec35c0cd5d1a1a8e7c8411db9ee29c2973).

For `compress`, that's what I originally tried but `ztd::Encoder` does not implement `std::io::Read`:

```
error[E0277]: the trait bound `zstd::Encoder<'_, &mut [u8]>: std::io::Read` is not satisfied
   --> tonic/src/codec/compression.rs:226:27
    |
226 |             std::io::copy(&mut zstd_encoder, &mut out_writer)?;
    |             ------------- ^^^^^^^^^^^^^^^^^ the trait `std::io::Read` is not implemented for `zstd::Encoder<'_, &mut [u8]>`
    |             |
    |             required by a bound introduced by this call
    |
```

Reply:
Thanks for the tip! Fixed in [0839324](https://github.com/hyperium/tonic/pull/1371/commits/08393243a2f1e7dc42dd53bf509ee915cb6f43aa)